File: secondary/tools/fdb/fdb_throughput.c
  // +build ignore

 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
 #include <stdint.h>
 #include <time.h>
 #include <sys/time.h>
 #if !defined(WIN32) && !defined(_WIN32)
 #include <unistd.h>
 #endif
 
 #include "libforestdb/forestdb.h"

 void do_test1()
 {

    printf("***** TEST1 SINGLE KV STORE 10M INSERT *****\n");

    fdb_file_handle *dbfile;
    fdb_kvs_handle *main, *back;

    char keybuf[256], bodybuf[256];
          
    fdb_config fconfig = fdb_get_default_config();
    fconfig.durability_opt = FDB_DRB_ASYNC;
    fdb_kvs_config kvs_config = fdb_get_default_kvs_config();

    fdb_open(&dbfile, "./ctest1", &fconfig);
    fdb_kvs_open(dbfile, &main, "main", &kvs_config);

    int i;
    int n = 10000000;

    char **keyarr = (char **)malloc(n * sizeof(char *)); 
    char **bodyarr = (char **)malloc(n * sizeof(char *)); 

    for (i=0;i<n;++i){
          keyarr[i] = (char *)malloc(25);
          bodyarr[i] = (char *)malloc(25);
          sprintf(keyarr[i], "key%d", i);
          sprintf(bodyarr[i], "body%d", i);
    }

    struct timeval ts_begin, ts_cur, ts_gap;
    gettimeofday(&ts_begin, NULL);
    for (i=0;i<n;++i){
          fdb_set_kv(main, (void*)keyarr[i], strlen(keyarr[i]), (void*)bodyarr[i], strlen(bodyarr[i]));
    }

    fdb_commit(dbfile, FDB_COMMIT_MANUAL_WAL_FLUSH);

    gettimeofday(&ts_cur, NULL);
    timersub(&ts_cur, &ts_begin, &ts_gap);

    printf("***** RESULT : Docs Inserted %d Time Taken %d secs\n", n, (int)ts_gap.tv_sec);

    fdb_kvs_close(main);
    fdb_close(dbfile);
    fdb_shutdown();
    system("rm ./ctest1");
}

 void do_test2()
 {

    printf("***** TEST2 TWO KV STORES 5M INSERT EACH *****\n");

    fdb_file_handle *dbfile;
    fdb_kvs_handle *main, *back;

    char keybuf[256], bodybuf[256];
          
    fdb_config fconfig = fdb_get_default_config();
    fconfig.durability_opt = FDB_DRB_ASYNC;
    fdb_kvs_config kvs_config = fdb_get_default_kvs_config();

    fdb_open(&dbfile, "./ctest2", &fconfig);
    fdb_kvs_open(dbfile, &main, "main", &kvs_config);
    fdb_kvs_open(dbfile, &back, "back", &kvs_config);

    int i;
    int n = 5000000;

    char **keyarr = (char **)malloc(n * sizeof(char *)); 
    char **bodyarr = (char **)malloc(n * sizeof(char *)); 

    for (i=0;i<n;++i){
          keyarr[i] = (char *)malloc(25);
          bodyarr[i] = (char *)malloc(25);
          sprintf(keyarr[i], "key%d", i);
          sprintf(bodyarr[i], "body%d", i);
    }

    struct timeval ts_begin, ts_cur, ts_gap;
    gettimeofday(&ts_begin, NULL);
    for (i=0;i<n;++i){
          fdb_set_kv(back, (void*)keyarr[i], strlen(keyarr[i]), (void*)bodyarr[i], strlen(bodyarr[i]));
          fdb_set_kv(main, (void*)keyarr[i], strlen(keyarr[i]), NULL, 0);
    }

    fdb_commit(dbfile, FDB_COMMIT_MANUAL_WAL_FLUSH);

    gettimeofday(&ts_cur, NULL);
    timersub(&ts_cur, &ts_begin, &ts_gap);
    printf("***** RESULT : Docs Inserted %d Time Taken %d secs\n", n, (int)ts_gap.tv_sec);

    fdb_kvs_close(main);
    fdb_kvs_close(back);
    fdb_close(dbfile);
    fdb_shutdown();
    system("rm ./ctest2");
}

 void do_test3()
 {

    printf("***** TEST3 TWO KV STORES 5M INSERT EACH WITH GET *****\n");

    fdb_file_handle *dbfile;
    fdb_kvs_handle *main, *back;

    char keybuf[256], bodybuf[256];
          
    fdb_config fconfig = fdb_get_default_config();
    fconfig.durability_opt = FDB_DRB_ASYNC;
    fdb_kvs_config kvs_config = fdb_get_default_kvs_config();

    fdb_open(&dbfile, "./ctest3", &fconfig);
    fdb_kvs_open(dbfile, &main, "main", &kvs_config);
    fdb_kvs_open(dbfile, &back, "back", &kvs_config);

    int i;
    int n = 5000000;
    size_t valuelen;
    void *value;

    char **keyarr = (char **)malloc(n * sizeof(char *)); 
    char **bodyarr = (char **)malloc(n * sizeof(char *)); 

    for (i=0;i<n;++i){
          keyarr[i] = (char *)malloc(25);
          bodyarr[i] = (char *)malloc(25);
          sprintf(keyarr[i], "key%d", i);
          sprintf(bodyarr[i], "body%d", i);
    }


    struct timeval ts_begin, ts_cur, ts_gap;
    gettimeofday(&ts_begin, NULL);
    for (i=0;i<n;++i){
          //fdb_doc_create(&doc[i], (void*)keybuf, strlen(keybuf),
           //      null, 0, (void*)bodybuf, strlen(bodybuf));
          //fdb_set(main, doc[i]);
          //fdb_doc_free(doc[i]);
          fdb_get_kv(back, (void*)keyarr[i], strlen(keyarr[i]), &value, &valuelen);
          fdb_set_kv(back, (void*)keyarr[i], strlen(keyarr[i]), (void*)bodyarr[i], strlen(bodyarr[i]));
          fdb_set_kv(main, (void*)keyarr[i], strlen(keyarr[i]), NULL, 0);
    }

    fdb_commit(dbfile, FDB_COMMIT_MANUAL_WAL_FLUSH);

    gettimeofday(&ts_cur, NULL);
    timersub(&ts_cur, &ts_begin, &ts_gap);
    printf("***** RESULT : Docs Inserted %d Time Taken %d secs\n", n, (int)ts_gap.tv_sec);

    fdb_kvs_close(main);
    fdb_kvs_close(back);
    fdb_close(dbfile);
    fdb_shutdown();
    system("rm ./ctest3");
}

int main(){
         do_test1();
         do_test2();
         do_test3();
}

File: secondary/tools/fdb/fdb_throughput.go
//go:build nolint

package main

import (
	"fmt"
	"log"
	"os"
	"time"

	forestdb "github.com/couchbase/indexing/secondary/fdb"
)

func main() {

	do_test1()
	do_test2()
	do_test3()

}

func do_test1() {

	log.Println("***** TEST1 SINGLE KV STORE 10M INSERT *****")

	config := forestdb.DefaultConfig()
	config.SetDurabilityOpt(forestdb.DRB_ASYNC)

	var dbfile *forestdb.File
	var err error
	file := "test1"
	if dbfile, err = forestdb.Open(file, config); err != nil {
		panic(err)
	}
	defer os.RemoveAll(file)
	defer forestdb.Destroy(file, config)

	var main *forestdb.KVStore
	kvconfig := forestdb.DefaultKVStoreConfig()
	if main, err = dbfile.OpenKVStore("main", kvconfig); err != nil {
		panic(err)
	}
	defer main.Close()

	log.Println("***** GENERATING *****")
	numItems := 10000000
	keys := make([][]byte, numItems)
	vals := make([][]byte, numItems)
	for i := 0; i < numItems; i++ {
		keys[i] = []byte(fmt.Sprintf("key%v", i))
		vals[i] = []byte(fmt.Sprintf("body%v", i))
	}

	time.Sleep(3 * time.Second)
	log.Println("***** LOADING*****")
	start := time.Now()
	for i := 0; i < numItems; i++ {
		main.SetKV(keys[i], vals[i])
	}
	dbfile.Commit(forestdb.COMMIT_MANUAL_WAL_FLUSH)
	elapsed := time.Since(start)

	log.Printf("***** RESULT : Docs Inserted %v Time Taken %v", numItems, elapsed)

}

func do_test2() {

	log.Println("***** TEST2 TWO KV STORES 5M INSERT EACH *****")

	config := forestdb.DefaultConfig()
	config.SetDurabilityOpt(forestdb.DRB_ASYNC)

	var dbfile *forestdb.File
	var err error
	file := "test2"
	if dbfile, err = forestdb.Open(file, config); err != nil {
		panic(err)
	}
	defer os.RemoveAll(file)
	defer forestdb.Destroy(file, config)

	var main, back *forestdb.KVStore
	kvconfig := forestdb.DefaultKVStoreConfig()
	if main, err = dbfile.OpenKVStore("main", kvconfig); err != nil {
		panic(err)
	}
	defer main.Close()

	if back, err = dbfile.OpenKVStore("back", kvconfig); err != nil {
		panic(err)
	}
	defer back.Close()

	log.Println("***** GENERATING *****")
	numItems := 5000000
	keys := make([][]byte, numItems)
	vals := make([][]byte, numItems)
	for i := 0; i < numItems; i++ {
		keys[i] = []byte(fmt.Sprintf("key%v", i))
		vals[i] = []byte(fmt.Sprintf("body%v", i))
	}

	time.Sleep(3 * time.Second)

	log.Println("***** LOADING*****")
	start := time.Now()
	for i := 0; i < numItems; i++ {
		back.SetKV(keys[i], vals[i])
		main.SetKV(keys[i], nil)
	}
	dbfile.Commit(forestdb.COMMIT_MANUAL_WAL_FLUSH)
	elapsed := time.Since(start)

	log.Printf("***** RESULT : Docs Inserted %v Time Taken %v", numItems, elapsed)

}

func do_test3() {

	log.Println("***** TEST3 TWO KV STORE 5M INSERT EACH WITH GET *****")

	config := forestdb.DefaultConfig()
	config.SetDurabilityOpt(forestdb.DRB_ASYNC)

	var dbfile *forestdb.File
	var err error
	file := "test3"
	if dbfile, err = forestdb.Open(file, config); err != nil {
		panic(err)
	}
	defer os.RemoveAll(file)
	defer forestdb.Destroy(file, config)

	var main, back *forestdb.KVStore
	kvconfig := forestdb.DefaultKVStoreConfig()
	if main, err = dbfile.OpenKVStore("main", kvconfig); err != nil {
		panic(err)
	}
	defer main.Close()

	if back, err = dbfile.OpenKVStore("back", kvconfig); err != nil {
		panic(err)
	}
	defer back.Close()

	log.Println("***** GENERATING *****")
	numItems := 5000000
	keys := make([][]byte, numItems)
	//lenkeys := make([]int, numItems)
	vals := make([][]byte, numItems)
	//lenvals := make([]int, numItems)
	for i := 0; i < numItems; i++ {
		keys[i] = []byte(fmt.Sprintf("key%v", i))
		//lenkeys[i] = len(keys[i])
		vals[i] = []byte(fmt.Sprintf("body%v", i))
		//	lenvals[i] = len(vals[i])
	}

	time.Sleep(3 * time.Second)
	log.Println("***** LOADING*****")
	start := time.Now()
	for i := 0; i < numItems; i++ {
		back.GetKV(keys[i])
		back.SetKV(keys[i], vals[i])
		//back.SetKV(keys[i], keys[i], lenkeys[i], lenkeys[i])
		main.SetKV(keys[i], nil)
	}
	dbfile.Commit(forestdb.COMMIT_MANUAL_WAL_FLUSH)
	elapsed := time.Since(start)
	log.Printf("***** RESULT : Docs Inserted %v Time Taken %v", numItems, elapsed)
}
